### ISSUE TITLE
`spawn` will not exit when finished if there are connections open

### ISSUE THREAD

**Issue by mikemaccana** [architect/sandbox#33]
**Describe the bug**
Found as part of architect/architect#354.

In a real AWS Lambda, [`context.callbackWaitsForEmptyEventLoop = false;`](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-context.html) can be set - this allows one to run a lambda without [disconnecting from database](https://www.mongodb.com/blog/post/optimizing-aws-lambda-performance-with-mongodb-atlas-and-nodejs). This is used to keep Lambda connections open on MongoDB, RethinkDB, etc.

Arc needs to simulate AWS's behavior locally. 

**We need to create the equivalent of `callbackWaitsForEmptyEventLoop` = false (ie, exit as soon as all processing is done, rather than waiting for the event loop to be empty?**

**To Reproduce**

Here is some demo code. Change `FUNCTION_NAME` to `'connectOnly'` or `'connectAndDisconnect'` to demonstrate the behavior.

Right now `'connectAndDisconnect'` works, but `'connectOnly'` does not - it finishes processing, but won't ever exit node since the connection is still open.

```
let spawn = require('child_process').spawn
var log = console.log.bind(console)
let minify = script=> '"' + script.replace(/\n/g, '').trim() + '"'

var functions = {
    connectAndDisconnect: `const rethinkdb = require('rethinkdb');    
    ;(async function(request, context) { 
        var connection = await rethinkdb.connect({ host: 'localhost', port: 28015 }); 
        await connection.close();
        console.log('I connected successfully!');
    })();`,
    connectOnly: `const rethinkdb = require('rethinkdb');    
    ;(async function(request, context) { 
        var connection = await rethinkdb.connect({ host: 'localhost', port: 28015 }); 
        console.log('I connected successfully!');
    })();`
}

const FUNCTION_NAME = 'connectOnly'

log(`Testing ${FUNCTION_NAME}`)

var functionToTest = functions[FUNCTION_NAME]

var child = spawn('node', ['-e', minify(functionToTest)], { 
    shell: true,
    stdio: [0, 'pipe', 'pipe'],
    cwd: 'C:\\Users\\mikem\\OneDrive\\Documents\\appless\\documentdb-arc-test\\src\\http\\get-index',
    env: {}
})

var completed = false;

let timeout = 5000;
// bake a timeout
setTimeout(function () {
    log(`Timed out after ${timeout} ms!`)
    child.kill()
}, timeout)

// 'close' will occur if the process finishes naturally
// 'exit' will occur if the process is killed
console.log(`What we want is to be see a 'close' event using the 'connectOnly' code above`)
;['close', 'exit'].forEach(function(event){
    child.on(event, function () {
        log(`'${event}' event happened!`)
        if ( ! completed ) {
            completed = true
            log(`Done!`)
        }
    })
})
```

**What happens?**

Unless we close the connection, the lambda will never complete and will time out

**Expected behavior**

Once the lambda has completed processing it should exit, even if the connection is still open.

**Comment by ryanblock** [comment]
Bumping this!

**Comment by mikemaccana** [comment]
Yeah I (or someone) still need to fix - it slows down the sandbox dramatically because the workaround is to open and close connections on every `spawn`. It also hit some folks in Slack https://architecture-as-text.slack.com/archives/C6BGT0D08/p1567294450054300

**Comment by jgallen23** [comment]
any workarounds here?  We are having this issue when trying to connect to mongo - basically prevents us from doing any local dev (which is a bit difficult)

**Comment by ryanblock** [comment]
I’d like to see this resolved, for sure. Right now I’m just not sure how with `spawn`. (That doesn’t mean it’s not possible, I’m just not sure how to make it happen.)

My sense is that emulating the event loop carryover between executions is not only going to be difficult, but is actually something of an anti pattern for local dev–right now. sandbox is highly deterministic due to its guaranteed statelessness between executions, and I think that’s an important characteristic for local dev, even if it sometimes comes at a minor performance penalty.

that said, just how big of a penalty is it to pay for closing and reopening mongo connections upon each execution?

Again, def open to ideas on how to ensure we automatically close connections when work is complete within spawn, please fire away!

**Comment by mikemaccana** [comment]
> Or is best thing to do something like if (process.env.NODE_ENV== 'testing') { db.close() } ?

> that said, just how big of a penalty is it to pay for closing and reopening mongo connections upon each execution?

@jgallen23 @ryanblock Good question! I've done that here, and it makes using the sandbox incredibly slow. I've also written a little Mongo-like adapter that writes to files (it's faster) but it's a pretty nasty hack. 

@ryanblock Would it be possible to just kill node (with a Unix signal) when we receive a response?

**Comment by ryanblock** [comment]
> Would it be possible to just kill node (with a Unix signal) when we receive a response?

I was thinking about something along those lines – and let Mongo deal with the stale connection cleanup on its own time. Will have an RC to test with this today.

**Comment by ryanblock** [comment]
While we’re on the topic, @mikemaccana @jgallen23 could one of y’all whip up a test case repo to test my fix against? Mongo instructions would be great too if you don’t mind!

**Comment by ryanblock** [comment]
Ok, fix for this is live in sandbox@RC! `npm i @architect/sandbox@RC` – please give it a whirl.

I'm happy with the changes, but I'm going to hold off merging the PR (https://github.com/architect/sandbox/pull/104) until we've verified the fix and put it through its paces.

**Comment by ryanblock** [comment]
Quick followup: build passes muster in some intense manual verification, and also clears my reduced test case (that uses `tail -f`), but will still stand by for additional verification around the original issue.

**Comment by mikemaccana** [comment]
Cool - have a couple off bugs in `deploy` to resolve right now but will check this out early this week.

**Comment by mikemaccana** [comment]
Using this now with Mongo/DocumentDB locally, having modified my code to not bother closing the connections. Seems faster. 🙂

**Comment by jgallen23** [comment]
The RC seems to work for me too

**Comment by ryanblock** [comment]
Right on! Will merge and release today. Please stand by.

**Comment by ryanblock** [comment]
Fix now live in Architect `6.0.15` and Sandbox `1.4.7`.
